### PR TITLE
Improve ISO-C compatability

### PR DESCRIPTION
**Pull request by TheOneric**
Does not actually change our target to ISO C99 + POSIX `str(n)dup`.  
~~I'm not sure about putting `ASS_PI` into `ass_compat.h`, but couldn't come up with a better place for it.~~ `ASS_PI` has been moved to `ass_utils.h`

### CODE REVIEW

File: libass/ass_render.c
Comment (MrSmile) [inline libass/ass_render.c:32]: I'm not a fan of excessive includes. `ass_render.c` already heavily depends on `ass_utils.h`, so adding one constant changes nothing.

File: libass/ass_render.c
Comment (MrSmile) [inline libass/ass_render.c:511]: I'd prefer symbols instead of letters here (`m_**`, `m_##`, `m_@@` or `m_$$`). It's not good to introduce new identifiers without definition, especially if that definition isn't quite trivial: `k, l = {x, y, z}`.

**Review by MrSmile — COMMENTED** [review]
Added a couple of comments.

**Review by astiob — APPROVED** [review]
LGTM.

(Also, I think you’ll enjoy https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/RTS.cpp#l148)

File: libass/ass_compat.h
Comment (TheOneric) [reply libass/ass_compat.h:30]: Added spaces, thx.
If with “inplace inline” you meant `static inline` function definitions inside `ass_compat.h`, I decided against that, because it results in identical two functions per source file being placed in the resulting lib *(if fallbacks are required)*. Also the "proper" `str(n)dup` functions aren't forced inline either; leaving this to the compiler and possibly LTO seems better overall imo.

File: libass/ass_render.c
Comment (TheOneric) [reply libass/ass_render.c:32]: I now see `ass_render.h` already includes `ass_utils.h`, which is definitely **not** where I'd expect this to be, as nothing in the header itself requires `ass_utils.h`.
Imo, a header file should only include the bare minimum it requires for the header to be usable, so if it eg declares a function returning `size_t`, it should include `<stddef.h>`, but *not* things only the associated source file needs. Each source file should include everything it uses directly, except for types etc inherited from its own header file.
This has the advantage of making implicit dependencies sneaking in less likely. If a header file removes an include, because its own source file no longer needs it, with implicit dependencies now suddenly *other* source files might break, because they relied on that implicit/transitive include. Also putting the includes for the *source* file in the *header* file, means a lot of unneeded garbage gets included in other source files including the header, which also results in unnecessary recompiles when *any* header gets edited.
Currently it seems libass' includes are just placed wherever, or is there some system to it?

… But this pr is not about cleaning up includes, so I just removed `ass_utils.h` from `ass_render.c` for now.

File: libass/ass_render.c
Comment (TheOneric) [reply libass/ass_render.c:511]: I find it harder to understand those two identical non-alphabetic symbols¹, instead of two different labels. Personally I'd assume any unbound label to be a placeholder for any valid value, especially for a common enumerators like`i, j, k, l`, but this might also be a non-universal convention like the `**, @@, …`.
Perhaps it's best to just add `for all k, l from {x, y, z}` to the first/all use(s)?

¹: In fact while context made the intent clear in the original `??` case, I thought the `??` would be used as shell globs, where `?` matches any single character.

Comment (astiob) [reply libass/ass_render.c]: Just to add another perspective:

> Personally I'd assume any unbound label to be a placeholder for any valid value, especially for a common enumerators like`i, j, k, l`

Personally, I’m not used to seeing enumerators go as far as ℓ (and k is borderline), so I’d be confused by a new, unexplained mention of something involving it. Even the new version with the explainer comment feels kinda weird to me.

I was going to say that I find it an improvement that this now explicitly describes exactly which index combinations are intended… but is that actually right here? meaningful? After another few glances at the surrounding text, this one sentence seems to mean to say:

> Let’s write D(thing) for the absolute error in the thing’s value that we’ll get after all transformations are done and reverted, relative to the value the thing would have in an ideal rendering. We will now find reduced expressions for D(various things) that involve only various combinations of D(any element of matrix m) and no D(other things).

which

1. isn’t immediately clear from the sentence at all, whether before or after this PR;
2. doesn’t necessary involve “all” combinations of x, y, z.

File: libass/ass_render.c
Comment (MrSmile) [reply libass/ass_render.c]: > I now see `ass_render.h` already includes `ass_utils.h`

It's also indirectly included dozen times through `ass_outline.h`.

> Currently it seems libass' includes are just placed wherever, or is there some system to it?

I've tried to minimize include count in [this PR](https://github.com/libass/libass/pull/430) but hadn't seen much enthusiasm about it.

File: libass/ass_render.c
Comment (MrSmile) [reply libass/ass_render.c]: > Personally, I’m not used to seeing enumerators go as far as ℓ

In real math texts there is always shortage of letters, so I've seen _ijklmn_ and even _pqrts_ as indices.
But in this particular case `m_ij` is indeed better than `m_kl` if we go with letters.

Comment (TheOneric) [reply libass/ass_render.c]: Now I've reworded the whole introductory paragraph, is this easier to understand and still correct?
@MrSmile: Why is `z0` a constant even though it depends on matrix coefficients? Also I don't understand what `ACCURACY` in the second touched paragraph is or why it is bit-anded with `D(y_out)`.

Comment (MrSmile) [reply libass/ass_render.c]: I think that rewording shouldn't be done in trigraph commit.

> `i, j` are already used as variables by the function.

It's unrelated context. Also mathematicians often reuse the same letters as indices (for example, index _i_ and imaginary unit _i_ in the same expression simultaneously).

> Why is `z0` a constant even though it depends on matrix coefficients?

It depends on initial non-quantized matrix coefficients. Also it's an arbitrary scale factor that cancels out later. So there are two equivalent approaches: divide all coefficients by `z0` first and consider `z0 = 1` later, or treat `z0` as constant.

> Also I don't understand what `ACCURACY` in the second touched paragraph is or why it is bit-anded with `D(y_out)`.

`ACCURACY` is target error, and all that text is in pseudo-math notation, where `&` is simply "and" and `~` is "have the same order as".

Comment (TheOneric) [reply libass/ass_render.c]: > I think that rewording shouldn't be done in trigraph commit.

Will split.

> > `i, j` are already used as variables by the function.
> 
> It's unrelated context. Also mathematicians often reuse the same letters as indices (for example, index _i_ and imaginary unit _i_ in the same expression simultaneously).

I've certainly seen that double use of `i` in the same expression before, but I always held the opinion, and still do, that this is poor style and should be avoided. In this case reusing `i, j` wouldn't be as bad, but we're not short on letters either, so using `k, l` is better imo and I don't see any advantage in using `i, j` instead.

> `ACCURACY` is target error, and all that text is in pseudo-math notation, where `&` is simply "and" and `~` is "have the same order as".

I'd rather use a comma instead of `&` and `≈` *(or just write out `same magnitude` if you want to stick to ASCII)* instead of `~`, which I *primarily* know as “proportional to”.

Comment (TheOneric) [reply libass/ass_render.c]: Later on you're also using `~=`, I assume this is supposed to be an ASCII-fallback for `≈`?

Comment (MrSmile) [reply libass/ass_render.c]: Actually, `i, j` have used in the code in exactly the same context: as vector/matrix indices. So reusing them is even better.

> which I _primarily_ know as “proportional to”.

It's not wrong too. Basically by `x ~ y` I mean `x ≈ C * y` with some constant not far from 1.

> Later on you're also using `~=`, I assume this is supposed to be an ASCII-fallback for `≈`?

Yea. Maybe we should decide that source encoding is strictly UTF-8 and permit special symbols? Math descriptions would be much better with them.

Comment (TheOneric) [reply libass/ass_render.c]: > Actually, i, j have used in the code in exactly the same context

It's used to enumerate various things, not only the matrix.

@MrSmile , @astiob : How's the new version?

File: libass/ass_render.c
Comment (astiob) [inline libass/ass_render.c:541]: English nit: “Note [A is B]” feels weird to me: I’d expect either “Note that A is B” or “Note: A is B” with a colon. Especially with “and that” following it, I think that leaves “Note that” as the only option.

Then again, I’m not actually a native speaker myself (so I may be wrong), and this paragraph has other (pre-existing) slight language problems (such as missing articles), and yet the meaning remains clear, so this isn’t super important.

**Review by astiob — APPROVED** [review]
I may have more questions about the mathematical _content_ of those paragraphs, but I accepted them when we first merged this (and I know I went through all the maths) and that’s not relevant to this PR. So I’ll refrain from any specific comments.

That leaves me with a few English-language nits, which I sincerely hope you don’t mind 😅 and other than that, it still looks good to me.

File: libass/ass_render.c
Comment (TheOneric) [reply libass/ass_render.c]: Not sure if it's necessary, but “Note *that* …” should be correct *(but in that case the second *that* seems superfluous)*; changed that and also added articles.

File: libass/ass_render.c
Comment (TheOneric) [reply libass/ass_render.c]: I often get confused with English punctuation; corrections are always welcome *(and are now applied, thanks)*.

File: libass/ass_render.c
Comment (MrSmile) [inline libass/ass_render.c:561]: If we permit unicode in source code then `<=` and `^2` should be replaced with `≤` and `²`, also better to use other symbols (`×`, `±`, `∝`, `≡`, `⋅`, `Δ`, `mᵢⱼ`, etc.) too.

File: libass/ass_render.c
Comment (MrSmile) [inline libass/ass_render.c:569]: Similarly, following expressions derived by substituting `ACCURACY` with `2 * POSITION_PRECISION`, it's not a reminder.

**Review by MrSmile — COMMENTED** [review]
Not sure what consensus about unicode in source is. Also `≈` is quite similar to `=`, contrary to `~=`, not sure if that is a problem.

Also I really think that better to use `m_ij` due to
- there is at least one person confused by `kl`;
- math notation uses `ij` with priority and collisions outside of index context don't matter;
- variables `i` and `j` have extensively used in relevant functions as index variables and there are no other conflicting uses.

File: libass/ass_utils.h
Comment (TheOneric) [reply libass/ass_utils.h]: The comment was copied from glibc's and musl's definition; now removed.

**Comment by TheOneric** [comment]
After going through this a bit more on IRC, I've now re-pushed a version of the rewording which should address everything from above, only has +3 lines and is imho easier to understand than the original version.

File: libass/ass_render.c
Comment (MrSmile) [inline libass/ass_render.c:543]: Later you've used `pick:`, should be uniform. Also, maybe `use` or `set` here will be better?

**Review by MrSmile — COMMENTED** [review]
Mostly LGTM, marked a couple of typos.

File: libass/ass_render.c
Comment (astiob) [reply libass/ass_render.c]: “Pick” conveys that the parameter is arbitrarily chosen to have this value. “Use” and “set” are more ambiguous. Frankly, I (still) don’t quite understand how `ACCURACY` is chosen or what it really is, and why that choice is OK, so “pick” would make me feel a little less lost.

**Comment by TheOneric** [comment]
Will merge this in two days if there are no objections.

**Pull request merged**
